haystack/finder.py

```diff
@@ -96,3 +98,164 @@ def get_answers_via_similar_questions(self, question: str, top_k_retriever: int
+
+        correct_retrievals = 0
+        correct_readings = 0
+        summed_avg_precision_retriever = 0
+        summed_avg_precision_reader = 0
+        exact_matches = 0
+        summed_f1 = 0
+        retrieve_times = []
+        read_times = []
+
+        # retrieve documents
+        questions_with_docs = []
+        retriever_start_time = time.time()
+        for q_idx, question in enumerate(questions):
+            question_string = question["_source"]["question"]
+            single_retrieve_start = time.time()
+            retrieved_docs = self.retriever.retrieve(question_string, top_k=top_k_retriever, index=doc_index)
+            retrieve_times.append(time.time() - single_retrieve_start)
+            for doc_idx, doc in enumerate(retrieved_docs):
+                # check if correct doc among retrieved docs
+                if doc.meta["doc_id"] == question["_source"]["doc_id"]:
+                    correct_retrievals += 1
+                    summed_avg_precision_retriever += 1 / (doc_idx + 1)
+                    questions_with_docs.append({
+                        "question": question,
+                        "docs": retrieved_docs,
+                        "correct_es_doc_id": doc.id})
+                    break
+        retriever_total_time = time.time() - retriever_start_time
+        number_of_questions = q_idx + 1
+
+        previous_top_k_per_candidate = self.reader.top_k_per_candidate
+        previous_return_no_answers = self.reader.return_no_answers
+        self.reader.top_k_per_candidate = 1
+        self.reader.return_no_answers = True
+        # extract answers
+        reader_start_time = time.time()
+        for q_idx, question in enumerate(questions_with_docs):
+            question_string = question["question"]["_source"]["question"]
+            docs = question["docs"]
+            single_reader_start = time.time()
+            predicted_answers = self.reader.predict(question_string, docs, top_k_reader)
+            read_times.append(time.time() - single_reader_start)
+            # check if question is answerable
+            if question["question"]["_source"]["answers"]:
+                for answer_idx, answer in enumerate(predicted_answers["answers"]):
+                    found_answer = False
+                    found_em = False
+                    best_f1 = 0
+                    # check if correct document
+                    if answer["document_id"] == question["correct_es_doc_id"]:
+                        summed_avg_precision_reader += 1 / (answer_idx + 1)
+                        gold_spans = [(gold_answer["answer_start"], gold_answer["answer_start"] + len(gold_answer["text"]) + 1)
+                                      for gold_answer in question["question"]["_source"]["answers"]]
+                        predicted_span = (answer["offset_start_in_doc"], answer["offset_end_in_doc"])
+
+                        for gold_span in gold_spans:
```

Comment:
This is a tough one and we should discuss more!
Computing EM and f1 is in other approaches only done on the best prediction. Should we also do that?
It becomes especially clear in the "no answer" case (further down) where any "no answer" prediction turns EM and f1 = 1.